### PR TITLE
Add indexes for CASCADE deletes for task_instance

### LINKED ISSUE #24484: `airflow db clean task_instance` takes a long time
### Apache Airflow version

2.3.1

### What happened

When I ran the `airflow db clean task_instance` command, it can take up to 9 hours to complete. The database around 3215220 rows in the `task_instance` table and  51602 rows in the `dag_run` table. The overall size of the database is around 1 TB.

I believe the issue is because of the cascade constraints on others tables as well as the lack of indexes on task_instance foreign keys. 

Running delete on a small number of rows gives this shows most of the time is spent in xcom and task_fail tables

```
explain (analyze,buffers,timing) delete from task_instance t1 where t1.run_id = 'manual__2022-05-11T01:09:05.856703+00:00'; rollback;
Trigger for constraint task_reschedule_ti_fkey: time=3.208 calls=23
Trigger for constraint task_map_task_instance_fkey: time=1.848 calls=23
Trigger for constraint xcom_task_instance_fkey: time=4457.779 calls=23
Trigger for constraint rtif_ti_fkey: time=3.135 calls=23
Trigger for constraint task_fail_ti_fkey: time=1164.183 calls=23
```

I temporarily fixed it by adding these indexes.

```
create index idx_task_reschedule_dr_fkey on task_reschedule (dag_id, run_id);
create index idx_xcom_ti_fkey on xcom (dag_id, task_id, run_id, map_index);
create index idx_task_fail_ti_fkey on task_fail (dag_id, task_id, run_id, map_index);
```

### What you think should happen instead

It should not take 9 hours to complete a clean up process. Before upgrading to 2.3.x, it was taking no more than 5 minutes.

### How to reproduce

_No response_

### Operating System

N/A

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dstandish**
When we add foreign keys with ON DELETE CASCADE, and we delete rows in the foreign table, the database needs to join back to the referencing table.  If there's no suitable index, then it can be slow to perform the deletes.

cc @wolfier 

closes #24484

### CODE REVIEW

**Review by blag — APPROVED** [review]

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+"""
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = 'f5fcbda3e651'
+down_revision = '3c94c427fdf6'
+branch_labels = None
+depends_on = None
+airflow_version = '2.4.0'
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
airflow_version = '2.3.3'
```

We should pull this into the next release.

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def upgrade():
+    """Apply Add indexes for CASCADE deletes"""
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
```

nit

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    with op.batch_alter_table('task_reschedule', schema=None) as batch_op:
+        batch_op.create_index('idx_task_reschedule_dag_run', ['dag_id', 'run_id'], unique=False)
+
+    with op.batch_alter_table('xcom', schema=None) as batch_op:
+        batch_op.create_index(
+            'idx_xcom_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index'], unique=False
+        )
+
+    # ### end Alembic commands ###
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
```

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def downgrade():
+    """Unapply Add indexes for CASCADE deletes"""
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
```

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    with op.batch_alter_table('xcom', schema=None) as batch_op:
+        batch_op.drop_index('idx_xcom_task_instance')
+
+    with op.batch_alter_table('task_reschedule', schema=None) as batch_op:
+        batch_op.drop_index('idx_task_reschedule_dag_run')
+
+    with op.batch_alter_table('task_fail', schema=None) as batch_op:
+        batch_op.drop_index('idx_task_fail_task_instance')
+
+    # ### end Alembic commands ###
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
```

File: airflow/models/taskfail.py [hunk 6]
```diff
@@ -39,6 +39,7 @@ class TaskFail(Base):
     duration = Column(Integer)
 
     __table_args__ = (
+        Index("idx_task_fail_task_instance", dag_id, task_id, run_id, map_index, unique=False),
```
Comment (jedcunningham) [inline airflow/models/taskfail.py]:
```suggestion
        Index("idx_task_fail_task_instance", dag_id, task_id, run_id, map_index),
```

That's the default, no?

File: airflow/models/taskreschedule.py [hunk 7]
```diff
@@ -61,6 +61,7 @@ class TaskReschedule(Base):
             name="task_reschedule_ti_fkey",
             ondelete="CASCADE",
         ),
+        Index('idx_task_reschedule_dag_run', dag_id, run_id, unique=False),
```
Comment (jedcunningham) [inline airflow/models/taskreschedule.py]:
```suggestion
        Index('idx_task_reschedule_dag_run', dag_id, run_id),
```

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+depends_on = None
+airflow_version = '2.4.0'
+
+
+def upgrade():
+    """Apply Add indexes for CASCADE deletes"""
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('task_fail', schema=None) as batch_op:
+        batch_op.create_index(
+            'idx_task_fail_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index'], unique=False
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
            'idx_task_fail_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index']
```

Same here.

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        )
+
+    with op.batch_alter_table('task_reschedule', schema=None) as batch_op:
+        batch_op.create_index('idx_task_reschedule_dag_run', ['dag_id', 'run_id'], unique=False)
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
        batch_op.create_index('idx_task_reschedule_dag_run', ['dag_id', 'run_id'])
```

File: airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    with op.batch_alter_table('xcom', schema=None) as batch_op:
+        batch_op.create_index(
+            'idx_xcom_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index'], unique=False
```
Comment (jedcunningham) [inline airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py]:
```suggestion
            'idx_xcom_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index']
```

**Review by jedcunningham — APPROVED** [review]

**Review by jedcunningham — APPROVED** [review]

**Review by jedcunningham — APPROVED** [review]

**Comment by calfzhou** [comment]
Upgrading airflow 2.3.2 to 2.3.3, db upgrade failed:

Running upgrade 3c94c427fdf6 -> f5fcbda3e651, Add indexes for CASCADE deletes on task_instance
...
MySQLdb.OperationalError: (1067, "Invalid default value for 'end_date'")

(MySQL 8.0)

**Comment by potiuk** [comment]
@calfzhou  - you might have your mysql configured with "NO ZERO DATE" and "STRICT MODE". 
https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_no_zero_date. 

Can you please confirm that? 

If so, you might want to disable it - https://stackoverflow.com/questions/9192027/invalid-default-value-for-create-date-timestamp-field.

**Comment by potiuk** [comment]
Cool! 

One question @calfzhou - maybe you would like to add a note about it as PR to https://airflow.apache.org/docs/apache-airflow/stable/howto/set-up-database.html?highlight=database+setup#setting-up-a-mysql-database  ? We already have a few watch-outs there and adding this one might be useful for others.

It's super easy - just click "Suggest a change on this page" and you will get a PR where you will be able to directly edit the page sources and you could just use similar approach as other warnings there ?

**Comment by calfzhou** [comment]
PR posted #24983

**Pull request merged**